src/benchmarklib/tpch/tpch_db_generator.cpp

```diff
@@ -0,0 +1,320 @@
+      table->add_column_definition(column_name_and_type[boost::hana::llong_c<0>],
+                                   column_name_and_type[boost::hana::llong_c<1>]);
+      return table;
+    });
+  }
+
+  std::shared_ptr<opossum::Table> finish_table() {
+    if (_current_chunk_row_count() > 0) {
+      _emit_chunk();
+    }
+
+    return _table;
+  }
+
+  void append_row(ColumnTypes&&... column_values) {
+    auto vectors_and_values = boost::hana::zip_with(
+        [](auto& vector, auto&& value) { return boost::hana::make_tuple(std::reference_wrapper(vector), value); },
+        _column_vectors, boost::hana::make_tuple(std::forward<ColumnTypes>(column_values)...));
+
+    boost::hana::for_each(vectors_and_values, [](auto vector_and_value) {
+      vector_and_value[boost::hana::llong_c<0>].get().push_back(vector_and_value[boost::hana::llong_c<1>]);
+    });
+
+    if (_current_chunk_row_count() >= _table->max_chunk_size()) {
+      _emit_chunk();
+    }
+  }
+
+ private:
+  std::shared_ptr<opossum::Table> _table;
+  boost::hana::tuple<opossum::pmr_concurrent_vector<ColumnTypes>...> _column_vectors;
+
+  size_t _current_chunk_row_count() const { return _column_vectors[boost::hana::llong_c<0>].size(); }
+
+  void _emit_chunk() {
+    opossum::Chunk chunk;
+
+    boost::hana::for_each(_column_vectors, [&](auto&& vector) {
+      using T = typename std::decay_t<decltype(vector)>::value_type;
+      chunk.add_column(std::make_shared<opossum::ValueColumn<T>>(std::move(vector)));
+      vector = typename std::decay_t<decltype(vector)>();
+    });
+    _table->emplace_chunk(std::move(chunk));
+  }
+};
+
+std::unordered_map<opossum::TpchTable, std::underlying_type_t<opossum::TpchTable>> tpch_table_to_dbgen_id = {
+    {opossum::TpchTable::Part, PART},     {opossum::TpchTable::PartSupp, PSUPP}, {opossum::TpchTable::Supplier, SUPP},
+    {opossum::TpchTable::Customer, CUST}, {opossum::TpchTable::Orders, ORDER},   {opossum::TpchTable::LineItem, LINE},
+    {opossum::TpchTable::Nation, NATION}, {opossum::TpchTable::Region, REGION}};
+
+template <typename DSSType, typename MKRetType, typename... Args>
+DSSType _call_dbgen_mk(size_t idx, MKRetType (*mk_fn)(DSS_HUGE, DSSType* val, Args...), opossum::TpchTable table,
+                       Args... args) {
+  /**
+   * Preserve calling scheme (row_start(); mk...(); row_stop(); as in dbgen's gen_tbl())
+   */
+
+  const auto dbgen_table_id = tpch_table_to_dbgen_id.at(table);
+
+  row_start(dbgen_table_id);
+
+  DSSType value;
+  mk_fn(idx, &value, std::forward<Args>(args)...);
+
+  row_stop(dbgen_table_id);
+
+  return value;
+}
+
+float _convert_money(DSS_HUGE cents) {
+  const auto dollars = cents / 100;
+  cents %= 100;
+  return dollars + (static_cast<float>(cents)) / 100.0f;
+}
+
+void _free_and_null_permutations(distribution* d) {
+  free(d->permute);
+  d->permute = NULL;
+}
+
+/**
+ * Call this after using dbgen to avoid memory leaks
+ */
+void _dbgen_cleanup() {
+  _free_and_null_permutations(&nations);
```

Comment:
`for (auto pointer : {&nations, ...}) _free_and_null_permutations(pointer)`?

Reply:
done